### PR TITLE
qtcreator: refactor for using qt54; qt4SDK, qt5SDK: commented

### PR DESCRIPTION
**Pull request by akaWolf**
Accordingly to #9342, this PR commented qt{4,5}SDK and refactor qtcreator.

TODO: properly documentation build.

### CODE REVIEW

**Comment by bjornfor** [comment]
If you're referring to the "withDocumentation = true" thingy, I have a fix for that in the pipeline: https://github.com/bjornfor/nixpkgs/commits/qt5-improvements

**Comment by akaWolf** [comment]
@bjornfor yeah, looks like https://github.com/bjornfor/nixpkgs/commit/6431ae0b76143549b768b2e93dbe4509de11decf should do the thing.

**Comment by akaWolf** [comment]
@bjornfor btw, did you try, documentation of Qt Creator package is really building with your fix?

**Comment by bjornfor** [comment]
@akaWolf: Yes, it works. It allows browsing the qtcreator documentation. What's missing is the qt dockumentation (classes, functions etc.). But it's been like this forever, I think.

I was thinking of pushing those commits to staging branch.

**Comment by akaWolf** [comment]
@bjornfor, but how passing -docdir parameter to Qt configure can interfere with Qt Creator building process?
"-docdir <dir> ......... Documentation will be installed to <dir> (default DATADIR/doc)"
Isn't it only change the Qt own install path?

**Comment by bjornfor** [comment]
As I commented in the commit, "qmake -query" returns an incorrect docdir if qt is not configured with explicit "-docdir ...". The qtcreator build system uses the output of "qmake -query" to look for docs. Hence the breakage.

**Comment by akaWolf** [comment]
@bjornfor well, then it should be merged! :)

cc @ttuegel

**Comment by akaWolf** [comment]
@bjornfor one more: how can wrong path of Qt doc (which can even not installed) affect the build process of Qt Creator?
are we talking about the same issue?

**Comment by bjornfor** [comment]
@akaWolf: I have no idea :-) I just debugged the failing withDocumentation=true build and found it was due to "qmake -query" returning something different from reality (regarding the docdir).

**Comment by akaWolf** [comment]
Well, it remains unclear for me, but good that it's working.

**Comment by akaWolf** [comment]
@bjornfor we need to merge your changes for Qt Creator be able automatically gets Qt docs.

**Comment by akaWolf** [comment]
since #9390 pushed, we can push that PR, and then add qtSDK packages.

**Comment by bjornfor** [comment]
@akaWolf: Is the need to wrap the qtcreator binary also a result of "qmake -query" mis-information?

**Comment by akaWolf** [comment]
@bjornfor no, it's the result of runtime only Qt plugin dependency.

**Comment by bjornfor** [comment]
@akaWolf: Are you sure there is no other way? I thought we were going for split qt + qtcreator packages, allowing users to pick their qt version at will, without rebuilding qtcreator. Ref. https://github.com/akaWolf/nixpkgs/commit/e921dec0d32b6fc70becd54664418317f8bbb5ba#commitcomment-12835868

CC @ttuegel

**Comment by akaWolf** [comment]
@bjornfor there is other way, if, as said @ttuegel, all env vars regarding Qt (to be more precise, QML2_IMPORT_PATH) are set appropriately, but I don't know, what @ttuegel want to do exactly.

**Comment by ttuegel** [comment]
@akaWolf I'm going to adjust the Qt `setupHook` to set those variables at build-time. Then all you will need to do is wrap the program with the environment variables provided. In the mean time, could you change the `buildInputs` for Qt Creator so that it only depends on the submodules it needs from `qt54`?

**Comment by akaWolf** [comment]
@ttuegel now buildInputs contains only modules, that are need for build + wrapper.

File: pkgs/development/qtcreator/default.nix [hunk 1]
```diff
@@ -54,6 +44,10 @@ stdenv.mkDerivation rec {
     Type=Application
     Categories=Qt;Development;IDE;
     __EOF__
+    # Wrap the qtcreator binary
+    wrapProgram $out/bin/qtcreator \
```
Comment (ttuegel) [inline pkgs/development/qtcreator/default.nix]: I just pushed the updated `setupHook` to master. Now you can wrap any Qt Creator (or any other Qt program) with:
```
wrapProgram $out/bin/qtcreator \
    --prefix QT_PLUGIN_PATH : "$QT_PLUGIN_PATH" \
    --prefix QML_IMPORT_PATH : "$QML_IMPORT_PATH" \
    --prefix QML2_IMPORT_PATH : "$QML2_IMPORT_PATH"
```
During the build, those variables will be populated with the appropriate values based on the dependencies of the program.

File: pkgs/top-level/all-packages.nix [hunk 2]
```diff
@@ -7945,10 +7945,10 @@ let
     developerBuild = true;
   });
 
-  qt4SDK = qtcreator.override {
+  /*qt4SDK = qtcreator.override {
```
Comment (ttuegel) [inline pkgs/top-level/all-packages.nix]: You can remove this entirely, since it will never work with Qt 4.

File: pkgs/top-level/all-packages.nix [hunk 3]
```diff
@@ -7969,15 +7969,14 @@ let
     developerBuild = true;
   });
 
-  qt5SDK = qtcreator.override {
+  /*qt5SDK = qtcreator.override {
```
Comment (ttuegel) [inline pkgs/top-level/all-packages.nix]: I think you can safely remove this now, too.

**Comment by akaWolf** [comment]
@ttuegel ok, I'll do some refactor of PR.

**Comment by bjornfor** [comment]
@akaWolf: Can you please update the commit message? For the changes it introduces, I think it deserves a bit more explanation. (And maybe split the commit in two?)

Do you have any comments on removing "QT_PRIVATE_HEADERS=${qtLib}/include"? (I'm trying to learn.)

**Comment by akaWolf** [comment]
Hm. @bjornfor 
```
Info: creating cache file /tmp/nix-build-qtcreator-3.4.2.drv-1/qt-creator-opensource-src-3.4.2/.qmake.cache
no configure script, doing nothing
building
make flags: SHELL=/nix/store/zmd4jk4db5lgxb8l93mhkvr3x92g2sx2-bash-4.3-p39/bin/bash    docs 
SRCDIR=/tmp/nix-build-qtcreator-3.4.2.drv-1/qt-creator-opensource-src-3.4.2/doc OUTDIR=/tmp/nix-build-qtcreator-3.4.2.drv-1/qt-creator-opensource-src-3.4.2/doc/html QTC_VERSION=3.4.2 QTC_VERSION_TAG=342 QTC_LICENSE_TYPE=opensource QT_INSTALL_DOCS=/tmp/nix-build-qtcreator-3.4.2.drv-1/qmake-qtcreator-3.4.2/doc QDOC_INDEX_DIR=/tmp/nix-build-qtcreator-3.4.2.drv-1/qmake-qtcreator-3.4.2/doc /tmp/nix-build-qtcreator-3.4.2.drv-1/qmake-qtcreator-3.4.2/bin/qdoc /tmp/nix-build-qtcreator-3.4.2.drv-1/qt-creator-opensource-src-3.4.2/doc/qtcreator.qdocconf
/tmp/nix-build-qtcreator-3.4.2.drv-1/qt-creator-opensource-src-3.4.2/doc/qtcreator.qdocconf:3: error: Cannot open file '/tmp/nix-build-qtcreator-3.4.2.drv-1/qmake-qtcreator-3.4.2/doc/global/qt-html-templates-offline.qdocconf': No such file or directory
Cannot open file '/tmp/nix-build-qtcreator-3.4.2.drv-1/qmake-qtcreator-3.4.2/doc/global/qt-html-templates-offline.qdocconf': No such file or directory

Aborting
Makefile:325: recipe for target 'html_docs' failed
make: *** [html_docs] Error 1
builder for ‘/nix/store/43z933n2b93q016nzbs1rmmqdq5vhjr9-qtcreator-3.4.2.drv’ failed with exit code 2
error: build of ‘/nix/store/43z933n2b93q016nzbs1rmmqdq5vhjr9-qtcreator-3.4.2.drv’ failed
```

**Comment by ttuegel** [comment]
@akaWolf I need to link `doc` in the `setupHook`, just a moment...

**Comment by ttuegel** [comment]
@akaWolf OK, the latest master has a patch that should fix that error.

**Comment by akaWolf** [comment]
@ttuegel 
```
strip:/nix/store/6iyd18y3pwsgjwsm03b2mqya1z1cdqs1-qtbase-5.4.2/lib/libQt5PlatformSupport.prl: File format not recognized
strip:/nix/store/6iyd18y3pwsgjwsm03b2mqya1z1cdqs1-qtbase-5.4.2/bin/syncqt.pl: File format not recognized
patching script interpreter paths in /nix/store/6iyd18y3pwsgjwsm03b2mqya1z1cdqs1-qtbase-5.4.2
/nix/store/6iyd18y3pwsgjwsm03b2mqya1z1cdqs1-qtbase-5.4.2/bin/syncqt.pl: interpreter directive changed from "/usr/bin/env perl" to "/nix/store/g33gb293p9azw9l5y14zii2xq7p8i104-perl-5.20.2/bin/perl "
/nix/store/6iyd18y3pwsgjwsm03b2mqya1z1cdqs1-qtbase-5.4.2/mkspecs/macx-ios-clang/rename_main.sh: interpreter directive changed from "/bin/bash" to "/nix/store/zmd4jk4db5lgxb8l93mhkvr3x92g2sx2-bash-4.3-p39/bin/bash  "
building path(s) ‘/nix/store/45qzxk7qfzkyrb7lli2x7xrbwwm2iy0s-qtactiveqt-5.4.2’
building path(s) ‘/nix/store/sr7qnwrsryacxjhfh8apgz08a76idp2m-qtsvg-5.4.2’
building path(s) ‘/nix/store/v61r1gqnd7gy51xsz1nbvn1rw29cc91q-qtxmlpatterns-5.4.2’
qmake: File exists
qmake: File exists
qmake: File exists
/nix/store/45qzxk7qfzkyrb7lli2x7xrbwwm2iy0s-qtactiveqt-5.4.2/share: No such file or directory
/nix/store/sr7qnwrsryacxjhfh8apgz08a76idp2m-qtsvg-5.4.2/share: No such file or directory
/nix/store/v61r1gqnd7gy51xsz1nbvn1rw29cc91q-qtxmlpatterns-5.4.2/share: No such file or directory
builder for ‘/nix/store/xdvkspg6q2rlwg2gp1jalv4dfs7psmrh-qtactiveqt-5.4.2.drv’ failed with exit code 1
cannot build derivation ‘/nix/store/5sm8jkjymhqzv0vscfi9jkfg415l91w9-qttools-5.4.2.drv’: 1 dependencies couldn't be built
cannot build derivation ‘/nix/store/zb0h3q9p5mahc0f46brfhdsz57yr9rls-qtscript-5.4.2.drv’: 1 dependencies couldn't be built
cannot build derivation ‘/nix/store/4kfkbrlsrp6gmsvm6wyha0vqxxb0lrnq-qtcreator-3.4.2.drv’: 1 dependencies couldn't be built
error: build of ‘/nix/store/4kfkbrlsrp6gmsvm6wyha0vqxxb0lrnq-qtcreator-3.4.2.drv’ failed
```

**Comment by ttuegel** [comment]
@akaWolf Oops, sorry! Pushed another patch, try it again with the latest master.

**Comment by akaWolf** [comment]
@ttuegel 
```
$ cat `which qtcreator`
#! /nix/store/zmd4jk4db5lgxb8l93mhkvr3x92g2sx2-bash-4.3-p39/bin/bash -e
export QT_PLUGIN_PATH=/nix/store/w523xdv73yrizaiscf9irkj5za3fkgs7-qtbase-5.4.2/lib/qt5/plugins:/nix/store/cm0p23p8sbrc0a0ba2zjcmmf3vwv64b9-qttools-5.4.2/lib/qt5/plugins:/nix/store/9qka6j1aq03mpy02xj5py10z2j5b0pk9-qtdeclarative-5.4.2/lib/qt5/plugins:/nix/store/8zz171fmapgaky9y85d22vij7h3jlvx7-qtsvg-5.4.2/lib/qt5/plugins:/nix/store/yl06byvlm4d1acha0bhp7fym2sizb37i-qtwebkit-5.4.2/lib/qt5/plugins:/nix/store/q1qr7wc6im6x9hprdpr4fw5z1ww1x77i-qtlocation-5.4.2/lib/qt5/plugins:/nix/store/d85jr7gsdx9qfmc3wixrxqj8hyfi8qrv-qtmultimedia-5.4.2/lib/qt5/plugins:/nix/store/6ghq1h8klylqvb2vl37710lvfklbj7cp-qtsensors-5.4.2/lib/qt5/plugins${QT_PLUGIN_PATH:+:}$QT_PLUGIN_PATH
export QML2_IMPORT_PATH=/nix/store/9qka6j1aq03mpy02xj5py10z2j5b0pk9-qtdeclarative-5.4.2/lib/qt5/qml:/nix/store/yl06byvlm4d1acha0bhp7fym2sizb37i-qtwebkit-5.4.2/lib/qt5/qml:/nix/store/q1qr7wc6im6x9hprdpr4fw5z1ww1x77i-qtlocation-5.4.2/lib/qt5/qml:/nix/store/d85jr7gsdx9qfmc3wixrxqj8hyfi8qrv-qtmultimedia-5.4.2/lib/qt5/qml:/nix/store/6ghq1h8klylqvb2vl37710lvfklbj7cp-qtsensors-5.4.2/lib/qt5/qml${QML2_IMPORT_PATH:+:}$QML2_IMPORT_PATH
exec /nix/store/zvbhp70b1frknpdiljl5hdi4whagrw23-qtcreator-3.4.2/bin/.qtcreator-wrapped "${extraFlagsArray[@]}" "$@"
```
where is Qt Declarative for QML2_IMPORT_PATH?
maybe something else is missing.

**Comment by akaWolf** [comment]
@bjornfor done.
In new Qt there is no need to provide private headers, since Qt modules are already provide that interface with public.

**Comment by akaWolf** [comment]
@ttuegel you can safely push that PR, since all work should be done at Qt package side now.

**Comment by bjornfor** [comment]
@akaWolf: Thanks. Can you also say "why" the SDKs are deleted (in the commit message).

**Comment by akaWolf** [comment]
@bjornfor yeah, I just refer to our discussion.

**Comment by bjornfor** [comment]
@akaWolf: IMHO, it'd be nice to summarize it in the commit message. (It's just one sentence, right?) Or both: summarize and link to discussion, if you want.

**Comment by akaWolf** [comment]
@bjornfor sorry, I'm unable to describe this shortly... :)
If you talk me what, I'll add it to commit msg.

**Comment by bjornfor** [comment]
@akaWolf: Well, I don't know _your_ reasons, but I can see a reason for removing qt4SDK; qtcreator doesn't support qt4 anymore:

```
$ nix-build -A qt4SDK
these derivations will be built:
  /nix/store/lg688yq1i1sy2vndv91647vnwsk8jccn-qtsdk-3.4.2-qt-4.8.7.drv
building path(s) ‘/nix/store/9cmxrd0138b84zv3gq2kqri3phcj8dxx-qtsdk-3.4.2-qt-4.8.7’
unpacking sources
unpacking source archive /nix/store/mvkk7yy6yqq31vhvhjj5h704d9a2qbf7-qt-creator-opensource-src-3.4.2.tar.gz
source root is qt-creator-opensource-src-3.4.2
patching sources
configuring
/tmp/nix-build-qtsdk-3.4.2-qt-4.8.7.drv-0/qt-creator-opensource-src-3.4.2/qtcreator.pri:73: Unknown replace function: clean_path
Project MESSAGE: Cannot build Qt Creator with Qt version 4.8.7.
Project ERROR: Use at least Qt 5.3.1.
```

Regarding qt5SDK, I'm not sure we should remove it yet.

Btw, it's your job as a PR submitter to tell others _why_ a change is needed ;-)

**Comment by akaWolf** [comment]
@bjornfor well, I need a @ttuegel's work to be done, anyway. :)

**Comment by ttuegel** [comment]
> Update 2: Ah, I see, I need to specify it at buildInputs. But that module isn't needed for build, just at runtime. Can/should we separate build- and run-time dependencies?
> 
> Update 3: Add to buildInputs of Quick Controls didn't help. Why there is also qttools, qtsvg, qtwebkit (!!!), qtlocation, qtmultimedia, qtsensors? We need only qtbase + qtscript + qtdeclarative for build and qtdeclarative + qtquickcontrols for runtime. Something is definitely broken.
> There are two ways: or we are set all qt submodules into env vars, or we set only that submodules, which we are setting in .nix file (for example, in buildInputs).
> Current situation does not fit into any of above ways.
> 
> Uptade 4: Anyway, we need there only those modules, which we are set in the .nix of package.
> The question is only: Should we split runtime and buildtime dependencies?

@akaWolf Nix has only build dependencies; there is no way to have separate run-time dependencies. There are extra packages that you didn't specify manually because any Qt package in `buildInputs` may have plugins or QML imports of its own that it needs at run-time. It is difficult to tell what different libraries may use internally. Even if these plugins aren't used by this program, it does no harm to include them, since they were already build dependencies.

**Comment by akaWolf** [comment]
@ttuegel I'm pretty sure, that Qt WebKit doesn't need for building another Qt modules... :)

**Comment by ttuegel** [comment]
@akaWolf Qt Tools and Qt Quick both have build dependencies on Qt Webkit.

**Comment by akaWolf** [comment]
@ttuegel maybe so, but I didn't specify none of them in 'buildInputs'.

**Comment by ttuegel** [comment]
@akaWolf Qt Script depends on Qt Tools.

**Comment by akaWolf** [comment]
@ttuegel But I have build Qt Creator few days ago with Qt 5.5.0 without Qt WebKit submodule.

**Comment by ttuegel** [comment]
@akaWolf Perhaps Qt Tools 5.5.0 does not depend on Qt Webkit, I haven't checked. But with Qt 5.4, you cannot depend on Qt Script without incurring a dependency on Webkit.

**Comment by akaWolf** [comment]
@ttuegel "--no-webkit" option has long existed, and always it was possible to build QtC. :)

**Comment by ttuegel** [comment]
@akaWolf If you think we should build with dependencies other than those listed in `qt.pro`, please open a separate issue/PR for that. Our Qt Tools are built with Qt Webkit, so the environment variable settings are correct.

**Comment by akaWolf** [comment]
@ttuegel I just have a look at Qt dependencies in another systems, and they are the same, as you described, so variables are correct, right.

**Pull request merged**